### PR TITLE
Adding Lock statements to ProgressPane Hide and Show functions.

### PR DESCRIPTION
**Pull request by pyrostew**
# PR Summary

Following the issue #17497 locks are being added to the Hide and Show methods in ProgressPane.cs to ensure that `_savedRegion` is not set to null after being check in the `IsShowing` property and being used in for loop in the Hide Method.

I am not sure if this is considered an acceptable fix for this issue, there might be a higher level that this would be better fixed at, but I'm not familiar enough with the code to know where that might be.

## PR Context
See Issue #17497 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None

### CODE REVIEW

**Comment by sandscap-sc** [comment]
@anmenaga @daxian-dbw @PaulHigin Can this be reviewed please? 

FYI - I received the following in VSCode debugging a script that was using a lot of progress bars in the PS Integrated Terminal.

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.PowerShell.ProgressPane.Hide()
   at Microsoft.PowerShell.ConsoleTextWriter.WriteLine(String value)
   at Microsoft.PowerShell.ConsoleHostUserInterface.WriteImpl(String value, Boolean newLine)
   at System.Management.Automation.MshCommandRuntime.WriteInformation(InformationRecord record, Boolean overrideInquire)
   at Microsoft.PowerShell.Commands.ReceiveJobCommand.WriteJobResults(Job job)
   at Microsoft.PowerShell.Commands.ReceiveJobCommand.WriteJobResultsRecursivelyHelper(Hashtable duplicate, Job job, Boolean registerInsteadOfWrite)
   at Microsoft.PowerShell.Commands.ReceiveJobCommand.WriteResultsForJobsInCollection(List`1 jobs, Boolean checkForRecurse, Boolean registerInsteadOfWrite)
   at Microsoft.PowerShell.Commands.ReceiveJobCommand.ProcessRecord()
   at System.Management.Automation.CommandProcessor.ProcessRecord()
```

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ProgressPane.cs [hunk 1]
```diff
@@ -26,7 +26,8 @@ class ProgressPane
         internal
         ProgressPane(ConsoleHostUserInterface ui)
         {
-            if (ui == null) throw new ArgumentNullException(nameof(ui));
+            if (ui == null)
+                throw new ArgumentNullException(nameof(ui));
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ProgressPane.cs:30]: Can't tell in this editor if it lines up right, but please add braces

```suggestion
               {
                   throw new ArgumentNullException(nameof(ui));
               }
```

**Review by SteveL-MSFT — APPROVED** [review]
Other than the 1 suggestion, the change looks fine to me

**Comment by daxian-dbw** [comment]
@SteveL-MSFT In what cases would `Hide` be called from multiple threads?

**Comment by iSazonov** [comment]
My doubts are there are already locks in some code paths and we can fall in dead lock. I'd want understand what is a code path for the PR issue which bypasses existing locks.

**Review by PaulHigin — COMMENTED** [review]
I don't know this code very well, but I am uncomfortable placing large sections of code within lock statements.  My concern is causing a regression where existing scripts start dead locking.

If the goal is to protect internal state (_savedRegion, _progressRegion), then I feel the fix should focus more narrowly on that.

**Comment by pyrostew** [comment]
I believe these locks are a pretty safe addition. The state of the class is fully internal, the locks could only be deadlocked if something outside of the class managed to get a lock on the lock object, but the lock object is private so that isn't possible. Sprinkling more smaller locking objects around just the internal state objects would increase the risk of mistakes and make it harder to maintain going forward. The objective is to only allow the one of the hide and show methods to be called at any one time. 

This was the simplest fix to an issue my team and I feel is pretty critical, clearly the ideal fix would be dig through the call tree and find where there is actually an issues with multiple threads interacting with the progress bars. but given the very specific scenario where I encountered this, it would be very complex to debug and find the true root cause.

**Comment by daxian-dbw** [comment]
I have the similar concern as @PaulHigin and @iSazonov.

By looking at the code, it seems the issue would happen only if more than one threads are calling `Hide` on the same `ProgressPane` instance -- one sets `_savedRegion` to null while the other is still at the early phase of the method.

The part that I don't understand is why there are multiple threads manipulating the same progress pane, and that feels to me is the root cause of this issue.

**Comment by iSazonov** [comment]
> The part that I don't understand is why there are multiple threads manipulating the same progress pane, and that feels to me is the root cause of this issue.

My understanding always was that ConsoleHast is _owner_. And it has already has locks to exclude race conditions.  But the issue shows there is a code path which bypasses these locks.

**Comment by pyrostew** [comment]
I'm happy to update this PR with suggestions for a more appropriate fix, but I'll need pointing at the right place as I'm not familiar with the architecture of the project.

**Comment by iSazonov** [comment]
> I'm happy to update this PR with suggestions for a more appropriate fix, but I'll need pointing at the right place as I'm not familiar with the architecture of the project.

You could look locks which works for local scenario. There are two places with such locks if I remember right. Then it would be great to find where the code is called in _remoting_ code. There is a _deserialization for ProgressRecord_ - a root of the issue there.

**Comment by PaulHigin** [comment]
The `WriteProgress` call from the remoting layer is likely through the `RemoteHost` object handling on the client.

https://github.com/PowerShell/PowerShell/blob/b780a34ad1a8b0fb4aa02c5a486d1e9a995615ab/src/System.Management.Automation/engine/remoting/common/WireDataFormat/RemoteHost.cs#L136

https://github.com/PowerShell/PowerShell/blob/b780a34ad1a8b0fb4aa02c5a486d1e9a995615ab/src/System.Management.Automation/engine/remoting/host/RemoteHostMethodInfo.cs#L136

The client normally marshals the remote host write onto the pipeline thread, and we should find out why that isn't happening in this case.

https://github.com/PowerShell/PowerShell/blob/b780a34ad1a8b0fb4aa02c5a486d1e9a995615ab/src/System.Management.Automation/engine/remoting/client/ClientMethodExecutor.cs#L71

**Comment by PaulHigin** [comment]
Looking at the original issue (#17497), I see that this only occurs during a WinRM remoting disconnect/reconnect event.  It appears that after reconnection, the remote host calls are no longer marshaled onto the client cmdlet thread.  I would like to first understand why this is happening before adding locks.  My guess is some client state information is lost after the reconnect.

**Comment by pyrostew** [comment]
Hello All, @TravisEz13 @adityapatwardhan 

Reawakening this one as I've just run headlong into it again on a newer version of PowerShell. I've rebased the original commit on master.

To make a point on some of the comments, it seems the feeling is that there is some issue further up the chain relative to the locks I've put in. I would argue that where I have put the locks in should have had them anyway. Whatever other issues exist, fundamentally there is a concurrent access issue in the ProgressPane class, and these changes fix that. If that means another bug becomes an non issue then isn't that win win?

I have had a go at trying to find the higher level issue but have found that I'm putting much more wide sweeping changes in to deal with it.

**Review by iSazonov — CHANGES_REQUESTED** [review]
I believe there must be at least two threads and most likely with remote sessions to reproduce the problem (one ends its progress bar, second continues its progress bar). In this case this fix is not completely correct.

**Pull request closed without merging**